### DISCUSSION TITLE
playground link - https://play.openpolicyagent.org/p/8cfHmyWx2J

### DISCUSSION THREAD

**Discussion by backbencherengineer** [open-policy-agent/community#336] [Community]
HI
i am trying use for loop, however its not working. 

# Loop over each key in mandatory_keys
for key in mandatory_keys {
  # Check if key is present in new_labels_keys
  if not key in new_labels_keys {
    all_keys_present = false
  }
}
}

label_pattern_match {

  pattern = "^[a-z0-9]+$"
  all_match := [regex.match(new_labels[key], pattern) | key := object.keys(new_labels)]
  all_match_result = all_match[_]
  all_match_result
}

values := {value | value := new_labels[_]}

**Comment by anderseknert** [comment]
Hi there 👋 There is no "for" construct in Rego. You can use `some .. in` for iteration, or `coll[_]`, etc... but in your example it looked like you wanted to check for `every`. I think this would do what you wanted: https://play.openpolicyagent.org/p/Yggru6R2zR

**Reply by backbencherengineer** [reply]
@anderseknert . Thank you for the solution. however below is an issue i found

the label_pattern_match condition is not working, because as per my current code I shouldn't have got the below message 
"deny": [
        "Mandatory labels do not match required pattern"
    ],
    
    because the pattern is correct for all the values present in the new_labels
    
    playground link - https://play.openpolicyagent.org/p/uqgirQz8cn

**Reply by anderseknert** [reply]
You had the `pattern` argument at the wrong place in the `regex.match` call. You'll probably want `every` there too https://play.openpolicyagent.org/p/VyglFIRCns

**Reply by backbencherengineer** [reply]
Thanks for the prompt response. let me test the solution

**Reply by backbencherengineer** [reply]
@anderseknert, 
I apologize for bothering you once more, but I have a question for you.

I have tested the pattern by beginning the value with numerical and then string (i.e new_labels := {"lifecycle": "123dev", "service": "123bb20", "cost": "123test123"}
but the output I got is     "label_pattern_match": true,, however, then it should result as false because our pattern is "^[a-z0-9]+$"  , which means the value should always start with a string and after then can have numbers 
https://play.openpolicyagent.org/p/0BDAoRhXZV

**Reply by charlieegan3** [reply]
> because our pattern is "^[a-z0-9]+$" , which means the value should always start with a string and after then can have numbers

That's not what this regex means. See here: https://regex101.com/r/yiXP6c/1

I think you're looking for a pattern like this: https://regex101.com/r/ILVixJ/1

**Reply by backbencherengineer** [reply]
thank you for your help, really appreciate it

**Reply by charlieegan3** [reply]
No probs, hope that works!

**Reply by backbencherengineer** [reply]
yes, it did.

**Reply by backbencherengineer** [reply]
Hi All,

I was trying to print the missing labels in sprintf section (i.e if any label is missing from mandatory_labels if should print the missing label)
and the same is with the pattern (i.e if patten is missing with any label values it should print the value for that)

i updated the code accordingly, however, i am not getting the desired result, can you please have a look?
rego playground - https://play.openpolicyagent.org/p/BwZx4k25l8

the error i am getting 
1 error occurred: policy.rego:12: rego_parse_error: unexpected ; token: expected "," or "]"
	msg := sprintf("Mandatory labels %v are missing", [key, _ in mandatory_labels; key not in new_labels_keys])

**Reply by charlieegan3** [reply]
Hi @backbencherengineer, I'm about to head out for the day but this might work for you: https://play.openpolicyagent.org/p/73isQqPRB3

**Reply by backbencherengineer** [reply]
Awesome. Thanks for all your help.
